lib/rspec-puppet/support.rb

```diff
@@ -217,7 +217,23 @@ def with_vardir
       end
     end
 
+    def reset_providers
+      Puppet::Type.eachtype do |type|
+        # Unset the default provider for each type as this is cached in a class
+        # instance method and so doesn't get re-evaluated when we change facts
+        # and compile a new catalogue.
+        type.defaultprovider = nil
+      end
+    end
+
     def build_catalog_without_cache(nodename, facts_val, trusted_facts_val, hiera_config_val, code, exported)
+      # Blindly stub out any feature and command Puppet::Confine evaluations
+      # that providers might do. This is necessary because we convert the
+      # resource catalogue to a RAL catalogue in order to evaluate all the
+      # automatic graph edges, which requires evaluating the providers.
+      allow_any_instance_of(Puppet::Confine::Feature).to receive(:pass?).with(anything).and_return(true)
+      allow_any_instance_of(Puppet::Util).to receive(:which).with(anything).and_return(true)
+      allow_any_instance_of(Puppet::Provider::Command).to receive(:execute).with(any_args).and_return("")
```

Comment:
> low-level stubs for Puppet::Util::Execution.execute, which will be skipped

Yeah, this stub could be trimmed down so that it's not so encompassing.

> As an alternative, maybe we can do it like e.g. stdlib and stub defaultprovider

It'd need some testing but I fear that this won't be that easy either. Stubbing `Puppet::Type#defaultprovider` with an existing type is easy enough, but what do we pick if the user hasn't specified a provider in their manifest and is expecting Puppet to choose the correct one? We could create a dummy provider, but then we'll need to stub out method in which the providers report their supported features back to the type otherwise we'll run into problems where type parameters become invalid because the provider doesn't support feature X (might not be that big of a problem, but it's not something that I've looked into before so I can't say for sure).